modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/_diffs/plugins/ballooneditor/plugin.js

```diff
@@ -277,111 +277,29 @@
 				);
 			};
 
+			const originalContextManagerCheck =
+				CKEDITOR.plugins.balloontoolbar.contextManager.prototype.check;
+
 			CKEDITOR.plugins.balloontoolbar.contextManager.prototype.check = function (
 				selection
 			) {
 				if (!selection) {
 					selection = this.editor.getSelection();
-
-					CKEDITOR.tools.array.forEach(
-						selection.getRanges(),
-						(range) => {
-							range.shrink(CKEDITOR.SHRINK_ELEMENT, true);
-						}
-					);
 				}
 
 				if (!selection) {
 					return;
 				}
 
-				const path = selection.getRanges()[0]?.startPath();
-
-				let contextMatched;
-
-				function matchEachContext(
-					contexts,
-					matchingFunction,
-					matchingArg1
-				) {
-					CKEDITOR.tools.array.forEach(contexts, (curContext) => {
-						if (
-							!contextMatched ||
-							contextMatched.options.priority >
-								curContext.options.priority
-						) {
-							const result = matchingFunction(
-								curContext,
-								matchingArg1
-							);
-
-							if (result instanceof CKEDITOR.dom.element) {
-								contextMatched = curContext;
-							}
-						}
-					});
-				}
-
-				function elementsMatcher(curContext, curElement) {
-					return curContext._matchElement(curElement);
-				}
-
-				matchEachContext(this._contexts, (curContext) => {
-					return curContext._matchRefresh(path, selection);
-				});
-
-				matchEachContext(this._contexts, (curContext) => {
-					return curContext._matchWidget();
-				});
-
-				if (path) {
-					const selectedElem = selection.getSelectedElement();
+				const selectedElement = selection.getSelectedElement();
```

Comment:
When double clicking an anchor element, a panel is shown and a text toolbar is being rendered behind.

Now(sorry for the distraction, I was testing when I decided to record this video):

https://user-images.githubusercontent.com/7663513/130659659-4908fa4d-3f8c-426d-9f51-355fe53ac9b7.mov

Before(without this refactor):

https://user-images.githubusercontent.com/7663513/130660090-92076382-0c18-4737-a95a-4f260034e4f5.mov

Is this behavior intended to be changed?

Reply:
> So, It's nice to remove every overriding code we have

It's been a while I had a look at this, but in this case if I remember correctly we weren't simply overriding, we were "copying existing code and changing a couple of lines" in that function.

I know this is rather "horrible" both to work with and maintain, but unfortunately the only way I found to modify that function so that the Balloon Editor behaved mostly like we wanted it to was this "hack and slash" way.

Reply:
Nice spot on bugs, @diegonvs ! See fixes in https://github.com/liferay-frontend/liferay-portal/pull/1395#issuecomment-905634997

About overriding prototypes. No, this is is not a good practice. It results in disjointed and flaky code. But the alternatives I can think of are dead ends:
- The correct OO way would be to extend class with custom behavior, or to write a new class. The problem is that all the standard plugins and UI elements would somehow need to know to use our new class, and this would be a massive change on base CKEditor.
- Another correct way would be to change base prototype directly, by making a patch on base repo. The problem is that we would be adding BE-specific logic to editor that needs to be generic.

As for overriding vs extending prototypes, this comes down to what makes sense. Extending is less invasive, hence the code improvement in this PR. But, this was only possible because all extended logic could have been done before (or after) original function. Sometimes, we need to override, if the required change is in the middle of the function. For example, positioning logic is completely different from base editor, so `CKEDITOR.ui.balloonPanel.prototype.attach` needs to be overriden.